### ISSUE TITLE
SniffNet Remote Agent to Monitor the Network Adapter on a Remote Host or Hosts

### ISSUE THREAD

**Issue by kenrmayfield** [GyulyVGC/sniffnet#303]
### Is there an existing issue for this?

- [X] I have searched the existing issues.

### Describe the solution you'd like

Setup a Remote Agent for SniffNet, just like Nagios's Remote Agent to Monitor the Network Adapter on a Remote Host or Hosts.
This would be a Option just like in SniffNet when you Pick which Network Adapter you would like to Monitor Network Traffic.
The Remote Agent would be Installed on Windows/Linux and Send Live Data of Network Traffic Information back to SniffNet.

**Comment by GyulyVGC** [comment]
As of now, the available adapters are those identified by pcap and are only internal to the machine on which Sniffnet is installed.

The idea you mentioned would require the development of the remote agent, if I understood correctly, and this is out of scope at the moment.

If you are aware, feel free to link some resource from which I can take inspiration for the modalities of work of remote agents and the transmission of data to the main application.

If the process wouldn't break too much the existing code base, I could consider this feature in the long term.

**Comment by kenrmayfield** [comment]
1. https://geekpeek.net/nagios-plugin-bash/
2. https://github.com/NagiosEnterprises/ncpa
3. https://github.com/NagiosEnterprises/nrpe
4. https://nsclient.org/
5. https://github.com/mickem/nscp

**Comment by GyulyVGC** [comment]
Thanks!
This seems a bit out of scope at the moment, since many other features are being worked on and none of them is going in the direction of a server + agent application, at least for the moment.

Feel free to leave this issue open, but I'll tag it as `wontfix` for now.

**Comment by aderusha** [comment]
> 1. https://geekpeek.net/nagios-plugin-bash/
> 2. https://github.com/NagiosEnterprises/ncpa
> 3. https://github.com/NagiosEnterprises/nrpe
> 4. https://nsclient.org/
> 5. https://github.com/mickem/nscp

I don't think any of these solutions will provide anything more than RX TX counters for each interface.  Nagios plugins aren't exporting port and IP src/dst addressing for each flow, it's just a count of how many packets came into and out of each individual network interface.

**Comment by kenrmayfield** [comment]
@GyulyVGC 

Have you thought about making a Agent for Windows and Daemon for Linux so that SniffNet can Sniff
Remote Machines?

**Comment by GyulyVGC** [comment]
> Have you thought about making a Agent for Windows and Daemon for Linux so that SniffNet can Sniff Remote Machines?

Hi @kenrmayfield, as I anticipated:

> This seems a bit out of scope at the moment, since many other features are being worked on and none of them is going in the direction of a server + agent application, at least for the moment.
Feel free to leave this issue open, but I'll tag it as wontfix for now.

I included this request in the [project's roadmap](https://github.com/GyulyVGC/sniffnet/blob/main/ROADMAP.md), but it'll be likely one of the last features to be implemented.

**Comment by GyulyVGC** [comment]
I'm seriously considering to introduce this feature in one of the next major releases of the app, probably version 1.6 or 1.7 (see the new [Roadmap](https://whimsical.com/sniffnet-roadmap-Damodrdfx22V9jGnpHSCGo)).

Starting to think about the implementation, I was exploring different possibilities for the protocol to use for data transmission between agent and Sniffnet.

A possible strategy that comes to my mind is to use gRPC, but I'm open to your suggestions.

**Comment by zero77** [comment]
@GyulyVGC 
Having an agent would be excellent but, it may be a lot of extra work and time to maintain it.

Have you thought about using an existing tool and just collecting the data from that. Then you wouldn't be responsible for maintaining an agent but ownley working with its data.

I am not sure if it would work but as an example there is opensnitch and it allows sending data to SIEM

https://github.com/evilsocket/opensnitch/wiki/SIEM-integration

**Comment by kenrmayfield** [comment]
@GyulyVGC 

Using a Existing Tool for Collecting Data is not a good idea for SniffNet.   If a User wants to Collect Data from Third Party then they would use something like Grafrana, InFluxDB and Telegraf.

You want your on STAMP on SnifNet without Third Party.   

A Agent specifically for SniffNet is the Best Way to Go.

**Comment by GyulyVGC** [comment]
Yeah, I'm in fact of the idea of developing a dedicated Sniffnet agent. 
While I understand that it'd introduce some maintenance efforts, I plan to keep it as a simple packet capture agent with limited configurations.

I just have to determine which is the best way to transmit data (e.g., gRPC, syslog, raw sockets, etc...).
By "best" I mean that we should satisfy the following:
- cross-platform
- efficient
- type-safe
- secure (integrity + encryption + authentication)
- minimal to no dependencies

**Comment by zero77** [comment]
if you want to develop your own agent, then i would go with your original comment of using gRPC

**Comment by kenrmayfield** [comment]
@GyulyVGC 

Remember when I stated to get Ideas from:

https://geekpeek.net/nagios-plugin-bash/
https://github.com/NagiosEnterprises/ncpa
https://github.com/NagiosEnterprises/nrpe
https://nsclient.org/
https://github.com/mickem/nscp

Also hey and how you been?

Thank You for still having this on the RoadMap and on Your Mind when I inquired about this 2 Years Ago.

**Comment by GyulyVGC** [comment]
Yeah I remember about them!
I'm fine, thanks 😄

**Comment by GyulyVGC** [comment]
Just adding a further observation to this discussion.

I'm not only planning to support a single remote host, but multiple of them.

This will require major backend & frontend changes to selectively enable or disable visualising traffic from one or more of the remote hosts.
In this way it'll be possible to monitor a group of systems and decide to periodically focus on a subset of them or even only a single one, without the need to restart the analysis.

**Comment by kenrmayfield** [comment]
@GyulyVGC 

This is what the Remote Agent allows for Multiple Systems to be Monitored.
Also like you stated.................No Need to Restart Analysis because the Remote Agent is Sending Data from Multiple Systems or Single System Actively Live which I am sure there will be a Database to Store the Data.

**Comment by GyulyVGC** [comment]
Hey everyone, I plan to start with the development of this feature in the coming weeks so this is one of the last chances to express feedbacks.

What I plan to develop is the following:
- a Sniffnet agent exporting network flows in IPFIX format
- new capabilities for the Sniffnet app itself to be an IPFIX collector, allowing to gather and show data from multiple IPFIX exporters simultaneously (either from Sniffnet agents or other IPFIX producers)

I often get requests about including support for OPNsense and OpenWRT, and I think such requests could be satisfied by this one.
Please correct me if I'm wrong and in such case feel free to share more about what you'd expect.

**Comment by zero77** [comment]
I don't know how the architecture will work yet, but having agent support for headerless servers would be great.
Maybe by having a web UI for headerless servers or sending logs to a device with a desktop UI, like OpenSnitch Firewall does.

**Comment by kenrmayfield** [comment]
@GyulyVGC 

How you been?

Excellent News to hear.

This is going to be a Great Edition to SniffNet.

I think this is going to push SniffNet even more over the top.

**Comment by GyulyVGC** [comment]
I just had an insightful conversation with the creator of [NetGauze](https://github.com/NetGauze/NetGauze) Ahmed Elhassany, and IPFIX seems to actually be suitable for what Sniffnet needs to implement this feature, so I'll move forward with the IPFIX-based implementation.

**Comment by GyulyVGC** [comment]
@all-contributors please add @ahassany for ideas.

**Comment by kenrmayfield** [comment]
@GyulyVGC 

Great News!!!!

**Comment by GyulyVGC** [comment]
A first step to finally close this issue is complete: I present you [`sniffnet-agent`](https://github.com/GyulyVGC/sniffnet-agent), the unit in charge to export traffic flows in IPFIX format.

Now what's missing is only the support for collecting IPFIX at Sniffnet-side, which I already started to work on in the `ipfix` branch.

**Comment by kenrmayfield** [comment]
@GyulyVGC 

How you been?

Good News about the SniffNet Agent.

**Comment by GyulyVGC** [comment]
And... boom! 💥 

This is now implemented thanks to PR #1270 
The feature is now on the `main` branch, and will be officially released in a couple months with Sniffnet v1.6.0.

Here is a preview:

https://github.com/user-attachments/assets/1cb9e3ae-9f43-4549-b778-70df5d3c7855
